src/python/WMCore/MicroService/MSPileup/MSPileupData.py

```diff
@@ -0,0 +1,177 @@
+    "currentRSEs": ["Disk1", "Disk3"],  # values provided by the MS itself
+    "fullReplicas": integer,  # total number of replicas to keep on Disk
+    "active": boolean,
+    "pileupSizeGB": integer, current size of the pileup in GB (integer)
+    "rulesList: list of strings (rules) used to lock the pileup id
+}
+
+The data flow should be done via list of objects, e.g.
+[<pileup object 1>, <pileup object 2>, ..., <pileup object n>]
+"""
+
+# third party modules
+from pymongo import IndexModel, errors
+
+# WMCore modules
+from WMCore.Database.MongoDB import MongoDB
+from WMCore.MicroService.MSCore.MSCore import MSCore
+
+
+def stripKeys(docs, stripKeys=['_id']):
+    """
+    Helper function to strip out keys from given dictionary. The input parameter
+    is a dictionary and keys are wiped out in place
+    :param docs: either input dictionary or list of dictionaries
+    :param stripKeys: list of of strip keys
+    """
+    if isinstance(docs, list):
+        results = []
+        for doc in docs:
+            if doc:
+                for key in stripKeys:
+                    doc.pop(key, None)
+                    results.append(doc)
+        return results
+
+    if docs and isinstance(docs, dict):
+        for key in stripKeys:
+            docs.pop(key, None)
+        return docs
+
+class MSPileupData(MSCore):
+    """
+    MSPileupData class provide whole logic behind data used and stored by MSPileup module
+    """
+
+    def __init__(self, msConfig, **kwargs):
+        super(MSPileupData, self).__init__(msConfig, **kwargs)
+
+        self.msConfig.setdefault("mongoDBRetryCount", 3)
+        self.msConfig.setdefault("mongoDBReplicaSet", None)
+        self.msConfig.setdefault("mongoDBPort", None)
+        self.msConfig.setdefault("mockMongoDB", False)
+
+        # A full set of valid database connection parameters can be found at:
+        # https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html
+        mongoIndex = IndexModel('pileupName', unique=True)
+        collection = self.msConfig.get('mongoDBCollection', 'msPileupCollection')
+        msOutDBConfig = {
+            'database': self.msConfig['mongoDB'],
+            'server': self.msConfig['mongoDBServer'],
+            'replicaSet': self.msConfig['mongoDBReplicaSet'],
+            'port': self.msConfig['mongoDBPort'],
+            'username': self.msConfig['mongoDBUser'],
+            'password': self.msConfig['mongoDBPassword'],
+            'mockMongoDB': self.msConfig['mockMongoDB'],
+            'connect': True,
+            'directConnection': False,
+            'logger': self.logger,
+            'create': True,
+            'collections': [(collection, mongoIndex)]
+        }
+        mongoDB = MongoDB(**msOutDBConfig)
+        self.msDB = getattr(mongoDB, self.msConfig['mongoDB'])
+        self.dbColl = self.msDB[collection]
+#         self.dbColl.create_indexes([mongoIndex])
```

Comment:
Following on what I've already written in here:https://github.com/dmwm/WMCore/issues/11424#issuecomment-1381935142
Everything in this `_init_` up to this point, I'd say needs to go in the service  and a single database session gets maintained per service poling cycle so we reuse resources. If we want to parallelize the queries to the database  there are other ways to do it, but not to create a single session with every document/object - if we go with a single document per PU this will increase a lot both memory footprint  and network usage.

Reply:
Todor, I disagree with your observation. Here is how I see separation of service and database layers with single session:
```
class MSPileupService():
    def __init__(self):
           # it keeps internal access to DB but it is re-used in this class across APIs, i.e. single session
           self.mspileup = MSPileupData()
    def postAPI(self, input):
          # it can call self.mspileup object to perform any database related operation, e.g.
          # create and store object
          cdict = validate(input)
          doc self.mspileup.create(cdict)
          self.mspileup.store(doc)
```
This way, you keep database details inside MSPileupData class and keep HTTP logic in Service class. There is no need to keep database logic within Service class. Each class responsible for its own behavior, i.e. Service class handles HTTP logic, while Data class handle database details.

Reply:
Well that's a matter of separation of code modules. All fine with me. I do not object having a separate file/module dealing with the database layer, as long as we have a single instance of it during the service execution and we  reuse the same object many times. I am actually more in favor of that -  this is  a sign of well organized code. Me likes! :)  

Then what I can sufggest to happen in this case is to create a third module, describing only the document schema and validation data logic, as done in the other micro services.  And this we cannot avoid, simply because there is a difference in the internal data interpretation between MongoDB and python - for this third one as described bellow we should have a single instance per pileup/object returned from the database.

Reply:
ok, then, we are in agreement.